### ISSUE TITLE
dangerous idiom in subsetting lesson

### ISSUE THREAD

**Issue by bbolker** [swcarpentry/r-novice-gapminder#286]
In the [subsetting lesson](https://github.com/swcarpentry/r-novice-gapminder/blob/gh-pages/_episodes_rmd/06-data-subsetting.Rmd), line 212, the text suggests the following approach to dropping an element with a specific name from a vector:
```
x[-which(names(x) == "a")]
```
This is a dangerous idiom to teach, because of a particular wart/infelicity in R. **If the name you're trying to drop doesn't exist, this will return an empty vector**, because `which()` will return `numeric(0)` (an empty numeric vector); negating `numeric(0)` returns `numeric(0)` again; and `x[numeric(0)]` is an empty vector.

One could argue that this is a teachable moment, but I think it risks teaching the wrong message ("R does scary things that are hard to predict that could break your workflow") and adds unnecessary cognitive load.

This problem is covered in Chapter 8 of Burns's [R Inferno](http://www.burns-stat.com/pages/Tutor/R_inferno.pdf) (p. 53); he says

> There are at least three possible solutions for the original problem.
```
out <- which(x2 == 5)
if(length(out)) x2[-out] else x2
```
> Another solution is to use logical subscripts:
```
x2[!(x2 %in% 5)]
```
> Or you can, in a sense, work backwards:
```
x2[ setdiff(seq_along(x2), which(x2 == 5)) ]
```

For the lesson I would suggest not using `which()` at all, and using logical indexing instead: `x[names(x)!="a"]` or `x[!(names(x) %in% c("a","b"))]`.  Happy to submit a pull request if people think this is the right way to go.

The only advantage I can see to the current treatment is that it covers a slightly more complex nested operation (logical -> integer via `which()` -> subset by index).  On the other hand, logical indexing is better here anyway ...

**Comment by dushoff** [comment]
I guess the problem is the negation trick, and the robust code closest in spirit to the bad code would be:

> x[which(names(x) != "a")]

Not that I have anything against any of the other solutions, but I'm surprised that one doesn't come up. The negation trick is the problem.

Another unmentioned solution is to keep the logic exactly the same by dropping the which:

> x[!(names(x) == "a")]

I'm aware that one of BB's solutions is very close to these).

**Comment by naupaka** [comment]
Thanks for pointing this out, @bbolker, and to @dushoff for the comments. My vote would be for 
```r
x[!(names(x) == "a")]
```
or 
```r
x[names(x) != "a"]
```
The second, which @bbolker suggested, seems the most readable to me, but I suppose it could go either way. Do either of you want to submit a quick PR to fix this?

**Comment by bbolker** [comment]
A few more issues here:

- while use of `which()` is listed as one of the episode objectives, this is the *only* use of `which()` in the episode. Making this example more sensible would require adding another example to illustrate the use of `which()`. I'm finding it challenging at the moment to think of a simple example where one would actually use `which(...)` *in subsetting* rather than using logical indexing directly (i.e., `x[which(cond)]` is almost always equivalent to `x[cond]` ...)
- this example also illustrates negative indexing.  I could write an example to illustrate negative indexing (e.g. `x[-1]` and `x[-length(x)]` are super-common pattern), but that will also add to the length of the episode.
- one of the challenges actually directly discusses this case (i.e., what happens with negative indexing when the name is missing, and how to debug it). That's actually OK, as long as we have previously introduced `which()` and negative indexing ...

These are all do-able, but I'm feeling nervous about making such big changes.  But I guess I should go ahead and suggest something which others can then reject or modify ... ?

**Comment by JCSzamosi** [comment]
If this is the only use of `which()` in the lesson, I don't think the lesson accomplishes the goal of teaching how to use it. My first inclination would be to take it out entirely. One can reasonably do without it (in this particular case, and in general) and the lessons are already really full.

**Comment by bbolker** [comment]
I looked at this some more, and it's going to require some more architectural changes - there are a sequence of concepts that need to be rearranged ...

**Comment by naupaka** [comment]
Closed by #290

**Comment by bbolker** [comment]
I can't re-open, but I think this needs to be revisited.  I'm getting desperately confused, so I'm not going to submit a pull request right now, but:

- at line 315 ("skipping named elements") the example is still screwed up; I *think* what should happen is that line 202 

```
x <- c(a=5.4, b=6.2, c=7.1, d=4.8, e=7.5) # we can name a vector 'on the fly'
```

should move to line 315.

(Also, there's not much consistency any more between single and double quotes - it looks like single quotes are in the majority ...)

**Comment by naupaka** [comment]
I just added that line again to fix it, see 0b3d5f63f12b391c17b4ba7753627b9d1ab55156 and a6b2792dbb26074ce90ef36ba0e6be93fef5b5ee. The problem was the `## Tip: Non-unique names` callout that changed what was stored in `x`. Not sure what happened...
